### ISSUE TITLE
Compatibility with Yarn 2 PnP?

### ISSUE THREAD

**Issue by neezer** [lukeed/uvu#133]
I can't get uvu working with [Yarn 2 PnP](https://next.yarnpkg.com/features/pnp). Here is an [example repo](https://github.com/neezer/uvu-yarn-pnp).

If I try `yarn uvu`, I get:

```
Error: ENOTDIR: not a directory, stat '/path/to/uvu-yarn-pnp/.yarn/cache/uvu-npm-0.5.1-ab55397928-f7bf150f1f.zip/node_modules/uvu/node_modules/uvu'
```

If I try `yarn uvu test.js`, I get:

```
Error: ENOTDIR: not a directory, scandir '/path/to/uvu-yarn-pnp/test.mjs'
```

If I try running the test script defined as `"scripts": { "test": "uvu" }`, I get:

```
Error: ENOTDIR: not a directory, stat '/path/to/uvu-yarn-pnp/.yarn/cache/uvu-npm-0.5.1-ab55397928-f7bf150f1f.zip/node_modules/uvu/node_modules/uvu'
```

If I try running the file directly with `node -r ./.pnp.cjs test.mjs`, I get:

```
node:internal/process/esm_loader:74
    internalBinding('errors').triggerUncaughtException(
                              ^

Error [ERR_MODULE_NOT_FOUND]: Cannot find package 'uvu' imported from /path/to/uvu-yarn-pnp/test.mjs
Did you mean to import uvu-yarn-pnp/.yarn/cache/uvu-npm-0.5.1-ab55397928-f7bf150f1f.zip/node_modules/uvu/dist/index.js?
    at new NodeError (node:internal/errors:371:5)
    at packageResolve (node:internal/modules/esm/resolve:712:9)
    at moduleResolve (node:internal/modules/esm/resolve:753:18)
    at Loader.defaultResolve [as _resolve] (node:internal/modules/esm/resolve:867:11)
    at Loader.resolve (node:internal/modules/esm/loader:89:40)
    at Loader.getModuleJob (node:internal/modules/esm/loader:242:28)
    at ModuleWrap.<anonymous> (node:internal/modules/esm/module_job:76:40)
    at link (node:internal/modules/esm/module_job:75:36) {
  code: 'ERR_MODULE_NOT_FOUND'
}
```

---

- Setting `pnpMode: loose` in my `.yarnrc.yml` file has no effect.
- `uvu` works fine if I use Yarn without PnP enabled (traditional `node_modules` folder).
- Using yarn 3.0.1

I ❤️ this library and want this to work. Please let me know if I can help diagnose further!

**Comment by lukeed** [comment]
Hi thanks, it should be fine? 

None of your cli usage looks correct -- please check out the [CLI docs](https://github.com/lukeed/uvu/blob/master/docs/cli.md) if you haven't already.

It will traverse everything, so by default, you're having it look at the project root and it's following nearly everything. Trying moving your test files to a "tests" directory and then do `uvu tests`

Hope that helps

**Comment by neezer** [comment]
Thanks for the speedy response! I made a `tests` directory and moved my `test.mjs` file into that directory, and renamed it `example.test.mjs`.

Then I ran `yarn uvu tests` and I get... the same thing as above:

```
Error: ENOTDIR: not a directory, stat '/path/to/uvu-yarn-pnp/.yarn/cache/uvu-npm-0.5.1-ab55397928-f7bf150f1f.zip/node_modules/uvu/node_modules/uvu'
```

I can't run `uvu` directly since the executable is not in my `$PATH`. I updated my `package.json` test script to read `uvu tests` and tried running `yarn test`... and got the same error.

Anything else I should try?

**Comment by neezer** [comment]
> None of your cli usage looks correct

I thought I could run an individual test file with `node` directly--per your docs on [Isolation](https://github.com/lukeed/uvu/blob/master/docs/cli.md#isolation)--which is what I thought I was doing with this line in my original post:

> ... `node -r ./.pnp.cjs test.mjs`

??

**Comment by neezer** [comment]
I forked your repo and am diving in to see if I can fix this, but in the meantime, here are is something else to consider:

With Yarn PnP turned off:

1. Clone the source code for `uvu` itself.
2. Install deps with `yarn`
3. Run `yarn test` (installs everything into `node_modules`)
4. See all 5 test suites pass.

With Yarn PnP turned on:

1. Clone the source code for `uvu` itself.
2. `yarn set version berry` (enable PnP)
3. Install deps with `yarn` (no `node_modules` folder anymore)
4. Run `yarn test`
5. See all 5 test suites fail; each suite has `"run w/o error code"` next to it.

I'll post other findings as I discover them.

**Comment by lukeed** [comment]
There shouldn't be any differences. uvu is just a plain old `bin` with all its dependencies defined correctly.

Looping in @arcanis and @paul-soporan who can hopefully shed some light here.

**Comment by arcanis** [comment]
My first thought is that might be related to PnP support of ESM (which we don't currently support, since it requires the experimental loader API and we're still [working on that](https://github.com/yarnpkg/berry/pull/2161)) 🤔

**Comment by lukeed** [comment]
Ok, thank you~! If that's the case, @neezer can you try something like `node -r esm tests/index.js` – you can still use ESM format in that file, but it'll be rewritten via the require hook. Notice the change from `.mjs` to `.js` extension.

If that works, then can you try `uvu -r esm tests` and see if that still works too?

This is just to rule out yarn@2's resolution algorithm and/or and symlinking issues.

If both work, then it's a stronger point towards @arcanis's ESM theory

**Comment by neezer** [comment]
>  @neezer can you try something like node -r esm tests/index.js – you can still use ESM format in that file, but it'll be rewritten via the require hook. Notice the change from .mjs to .js extension.

Modifying your command to include PnP thusly:

```
yarn node -r esm tests/index.js
```

Sadly, that gives me a whole heap of minified JS code to stdout, and then this at the bottom:

```
TypeError: Ko._resolveFilename is not a function
    at node:internal/main/run_main_module:17:47
```

**Comment by neezer** [comment]
I also tried rewriting my test file to CommonJS, and I got the same `ENOTDIR` errors from above.

Eg., `tests/example.test.js`:

```js
const { test } = require('uvu');
const assert = require('uvu/assert');

test('Math.sqrt()', () => {
  assert.is(Math.sqrt(4), 2);
  assert.is(Math.sqrt(144), 12);
  assert.is(Math.sqrt(2), Math.SQRT2);
});

test.run();
```

... then running with `yarn uvu tests`, which yielded

```
Error: ENOTDIR: not a directory, stat '/path/to/my/project/root/uvu-yarn-pnp/.yarn/cache/uvu-npm-0.5.1-ab55397928-f7bf150f1f.zip/node_modules/uvu/node_modules/uvu'
    at statSync (node:fs:1526:3)
    at tryStatSync (node:internal/modules/esm/resolve:122:13)
    at packageResolve (node:internal/modules/esm/resolve:687:18)
    at moduleResolve (node:internal/modules/esm/resolve:753:18)
    at Loader.defaultResolve [as _resolve] (node:internal/modules/esm/resolve:867:11)
    at Loader.resolve (node:internal/modules/esm/loader:89:40)
    at Loader.getModuleJob (node:internal/modules/esm/loader:242:28)
    at Loader.import (node:internal/modules/esm/loader:177:28)
    at importModuleDynamically (node:internal/modules/cjs/loader:1041:29)
    at importModuleDynamicallyWrapper (node:internal/vm/module:437:21)
```

---

Is there something about how `uvu` loads itself/test suites that may be tripping this up? I see in `bin.js` the `hasImport` check, which I imagine would probably return true on newer versions of NodeJS regardless if the users tests were actually written with ES Modules... could that be forcing the use of `import`/`export` syntax, which isn't yet compatible with PnP?

**Comment by neezer** [comment]
Ok, check this change for [bin.js](https://github.com/lukeed/uvu/blob/master/bin.js#L8-L11) out:

```js
const hasImport = (() => {
  if (process.versions && process.versions.pnp) return false; // <-- addition
  try { new Function('import').call(0) }
  catch (err) { return !/unexpected/i.test(err.message) }
})();
```

Per https://yarnpkg.com/advanced/pnpapi#processversionspnp:

> When operating under PnP environments, this value will be set to a number indicating the version of the PnP standard in use (which is strictly identical to require('pnpapi').VERSIONS.std).

I believe this change forces `uvu` to use `require` (which PnP can hook into) instead of `import`/`export` (which is available on newer Node versions, and so would not throw in the following `try`/`catch`).

This works when my test is written in CommonJS (`uvu` **_can_** run my tests with PnP enabled), but still breaks when my test is written with ESM (with or without the presence of `-r esm`). However, the error is much clearer, as now Yarn gives me this:

```
Error [ERR_REQUIRE_ESM]: require() of ES Module /path/to/code/uvu-yarn-pnp/tests/example.test.mjs not supported.
Instead change the require of /path/to/code/uvu-yarn-pnp/tests/example.test.mjs to a dynamic import() which is available in all CommonJS modules.
```

**Comment by lukeed** [comment]
So all of those examples work in a plain npm/pnpm/yarn@1 setup – with the exception of `-r esm example.test.mjs` as the `esm` will not touch `.mjs` files; which means you've told uvu to enter CommonJS mode yet it's trying to require() ESM test code.

> I see in bin.js the hasImport check, which I imagine would probably return true on newer versions of NodeJS regardless if the users tests were actually written with ES Modules

@neezer the `hasImport` check just sees in the current Node version supports `import`. If it does and _if there were not `-r/--require` flags_, then it will be loading tests via `import()` which works with tests in ESM **or** CommonJS format. However, if you use `-r/--require`, then tests are always loaded via `require()`, regardless of what your Node version supports. I need to be more clear, please try exactly this:

```js
// package.json
{
  "scripts": {
    "test": "uvu -r esm tests"
  }
}

// tests/index.js
import { test } from 'uvu';
import * as assert from 'uvu/assert';

test('Math.sqrt()', () => {
  assert.is(Math.sqrt(4), 2);
  assert.is(Math.sqrt(144), 12);
  assert.is(Math.sqrt(2), Math.SQRT2);
});

test.run();
```

If this also fails, then this is an issue with yarn@2 resolving here, on top of any ESM problem.

**Comment by neezer** [comment]
[Updated](https://github.com/neezer/uvu-yarn-pnp) as you suggested. 

Running `yarn test` yields:

```
uvu-yarn-pnp on main via 6.6.2 took 2s
❯ yarn test
/path/to/code/uvu-yarn-pnp/tests/index.js:1
(function (exports, require, module, __filename, __dirname) { import { test } from 'uvu';
                                                              ^^^^^^

SyntaxError: Cannot use import statement outside a module
    at new Script (node:vm:100:7)
    at createScript (node:vm:257:10)
    at Object.runInThisContext (node:vm:305:10)
    at wrapSafe (node:internal/modules/cjs/loader:1020:15)
    at Module._compile (node:internal/modules/cjs/loader:1065:27)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1153:10)
    at Module.load (node:internal/modules/cjs/loader:981:32)
    at Function.external_module_.Module._load (/path/to/code/uvu-yarn-pnp/.pnp.cjs:9533:14)
    at Module.require (node:internal/modules/cjs/loader:1005:19)
    at require (node:internal/modules/cjs/helpers:94:18)
```

---

> ... as the esm will not touch .mjs files

My understanding is that [esm does have basic support](https://www.npmjs.com/package/esm#features):

> 🔒 .mjs files are limited to basic functionality without support for esm options.

**Comment by lukeed** [comment]
It does not. `esm` will not touch `.mjs` files. Please repro exactly as shown above.

**Comment by lukeed** [comment]
This is definitely a yarn-specific issue. uvu is doing everything correctly & plays well with every other package manager. I've seen these types of issues pop up in other places _specifically_ with yarn@2+, so there's nothing to be done here.

If you take your repo as it is now, and use yarn@1 or npm or pnpm, it'll work as expected.

**Comment by neezer** [comment]
> This is definitely a yarn-specific issue.

Agreed.

Thank you for your time & attention to this issue! Huge fan of your work. 🥇 

---

**TL;DR** For anyone who scrolls to the bottom of this issue looking for a definitive answer:

- [uvu](https://github.com/lukeed/uvu) + [Yarn PnP](https://yarnpkg.com/features/pnp) = 💥 (until Yarn PnP [supports ESM](https://github.com/yarnpkg/berry/pull/2161))
- [esm](https://www.npmjs.com/package/esm) won't help

**Comment by azat-io** [comment]
Is there any way to run my Typescript test with Yarn berry?

I tried to run tests using the command `yarn tsm node_modules/uvu/bin.js .` but got error:

```
yarn tsm node_modules/uvu/bin.js . \\.test\\.ts
(node:58882) ExperimentalWarning: --experimental-loader is an experimental feature. This feature could change at any time
(Use `node --trace-warnings ...` to show where the warning was created)
(node:58883) ExperimentalWarning: --experimental-loader is an experimental feature. This feature could change at any time
(Use `node --trace-warnings ...` to show where the warning was created)
fs.js:1131
  handleErrorFromBinding(ctx);
  ^

Error: ENOTDIR: not a directory, stat '/Users/user/Developer/cpa-auth/.yarn/cache/tsm-npm-2.2.1-7322a65949-ec9c0c59f0.zip/node_modules/tsm/loader.mjs'
```
